[ADT] SmallSet is unusable in make_filter_range due to improper iterator facade usage

On a rather simple testcase attempt to iterate through "filtered" SmallSet fails on iterator dereference.
It tries to bind non-const reference to const value.

```
] cat test.cc
#include <llvm/ADT/SmallSet.h>
#include <llvm/ADT/STLExtras.h>
void test() {
  llvm::SmallSet<int, 12> NewSet;
  auto PredTrue = [](int A) { return true; };
  auto Filtered = llvm::make_filter_range(NewSet, PredTrue);
  for (const auto &Val : Filtered) {
    (void)Val;
  }
}
] gcc test.cc -I../llvm/include -I./include -std=c++17  
In file included from ../llvm/include/llvm/ADT/SmallSet.h:21,
                 from test.cc:1:
../llvm/include/llvm/ADT/iterator.h: In instantiation of ‘ReferenceT llvm::iterator_adaptor_base<DerivedT, WrappedIteratorT, IteratorCategoryT, T, DifferenceTypeT, PointerT, ReferenceT>::operator*() const [with DerivedT = llvm::filter_iterator_base<llvm::SmallSetIterator<int, 12, std::less<int> >, test()::<lambda(int)>, std::forward_iterator_tag>; WrappedIteratorT = llvm::SmallSetIterator<int, 12, std::less<int> >; IteratorCategoryT = std::forward_iterator_tag; T = int; DifferenceTypeT = long int; PointerT = int*; ReferenceT = int&]’:
../llvm/include/llvm/ADT/STLExtras.h:471:28:   required from ‘decltype(auto) llvm::filter_iterator_base<WrappedIteratorT, PredicateT, IterTag>::operator*() const [with WrappedIteratorT = llvm::SmallSetIterator<int, 12, std::less<int> >; PredicateT = test()::<lambda(int)>; IterTag = std::forward_iterator_tag]’
  471 |     return BaseT::operator*();
      |            ~~~~~~~~~~~~~~~~^~
test.cc:7:26:   required from here
    7 |   for (const auto &Val : Filtered) {
      |                          ^~~~~~~~
../llvm/include/llvm/ADT/iterator.h:305:42: error: binding reference of type ‘int&’ to ‘const int’ discards qualifiers
  305 |   ReferenceT operator*() const { return *I; }
      |                                          ^
```

This happens because SmallSetIterator fails to define its reference type as "const".
SmallSet does have const iterator by design, so SmallSetIterator should set ReferenceT parameter of its base class iterator_facade_base to "const T&".

I dont see an easy way to work around this issue.

--- a/llvm/include/llvm/ADT/SmallSet.h
+++ b/llvm/include/llvm/ADT/SmallSet.h
@@ -32,7 +32,8 @@ namespace llvm {
 template <typename T, unsigned N, typename C>
 class SmallSetIterator
     : public iterator_facade_base<SmallSetIterator<T, N, C>,
-                                  std::forward_iterator_tag, T> {
+                                  std::forward_iterator_tag, T, std::ptrdiff_t,
+                                  const T *, const T &> {
 private:
   using SetIterTy = typename std::set<T, C>::const_iterator;
   using VecIterTy = typename SmallVector<T, N>::const_iterator;

--- a/llvm/unittests/ADT/SmallSetTest.cpp
+++ b/llvm/unittests/ADT/SmallSetTest.cpp
@@ -15,6 +15,7 @@
 #include "gmock/gmock.h"
 #include "gtest/gtest.h"
 #include <string>
+#include <type_traits>
 
 using namespace llvm;
 
@@ -251,6 +252,22 @@ TEST(SmallSetTest, IteratorIncMoveCopy) {
   EXPECT_EQ("str 0", *Iter);
 }
 
+template <typename T>
+constexpr bool is_const_ref = std::is_const_v<std::remove_reference_t<T>>;
+
+TEST(SmallSetTest, IteratorDerefConst) {
+  // Verify that dereference of SmallSet's iterator gives const-reference.
+  SmallSet<int, 4> sint;
+  EXPECT_TRUE(is_const_ref<decltype(sint)::const_iterator::reference>);
+  EXPECT_TRUE(is_const_ref<decltype(*sint.begin())> &&
+              is_const_ref<decltype(*sint.end())>);
+
+  SmallSet<std::string, 4> sstr;
+  EXPECT_TRUE(is_const_ref<decltype(sstr)::const_iterator::reference>);
+  EXPECT_TRUE(is_const_ref<decltype(*sstr.begin())> &&
+              is_const_ref<decltype(*sstr.end())>);
+}
+
 TEST(SmallSetTest, EqualityComparisonTest) {
   SmallSet<int, 8> s1small;
   SmallSet<int, 10> s2small;
@@ -299,4 +316,15 @@ TEST(SmallSetTest, Contains) {
   EXPECT_TRUE(Set.contains(0));
   EXPECT_TRUE(Set.contains(1));
   EXPECT_TRUE(Set.contains(2));
+}
+
+TEST(SmallSetTest, FilterPositive) {
+  SmallSet<int, 4> Set({3, -1, 2, -4});
+  auto PositiveRange = make_filter_range(Set, [](int V) { return V > 0; });
+  auto PositiveIt = PositiveRange.begin();
+  EXPECT_EQ(*PositiveIt, 3);
+  ++PositiveIt;
+  EXPECT_EQ(*PositiveIt, 2);
+  ++PositiveIt;
+  EXPECT_EQ(PositiveIt, PositiveRange.end());
 }

